### ISSUE TITLE
sql: Support view queries with star expansions

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#10028]
Copied over from https://github.com/cockroachdb/cockroach/issues/9921#issuecomment-254009800

We currently don't expand out star selectors before persisting a new view's query. If a dependent table has columns added to it, that can break later queries from the view depending on how that dependent table was used.

It's kind of a fluke, but a simple star select works just fine as columns are added because the view descriptor only lists the original columns as its ResultColumns:

```
root@:26257> create table star (a int primary key, b int);
CREATE TABLE
root@:26257> create view v as select * from star;
CREATE VIEW
root@:26257> insert into star values (1, 1), (2, 2);
INSERT 2
root@:26257> select * from star;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
| 2 | 2 |
+---+---+
(2 rows)
root@:26257> SELECT * FROM v;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
| 2 | 2 |
+---+---+
(2 rows)
root@:26257> alter table star add column c int;
ALTER TABLE
root@:26257> SELECT * FROM star;
+---+---+------+
| a | b |  c   |
+---+---+------+
| 1 | 1 | NULL |
| 2 | 2 | NULL |
+---+---+------+
(2 rows)
root@:26257> SELECT * FROM v;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
| 2 | 2 |
+---+---+
(2 rows)
```

On the other hand, if the \* isn't at the end of the query, it can break the view as columns are added:

```
root@:26257> create table t (a int primary key, b int);
CREATE TABLE
root@:26257> insert into t values (1, 1), (2, 2);
INSERT 2
root@:26257> CREATE VIEW tv AS SELECT t1.*, t2.b AS t2b FROM t AS t1 JOIN t AS t2 ON t1.a = t2.a;
CREATE VIEW
root@:26257> select * from tv;
+---+---+-----+
| a | b | t2b |
+---+---+-----+
| 1 | 1 |   1 |
| 2 | 2 |   2 |
+---+---+-----+
(2 rows)
root@:26257> ALTER TABLE t ADD COLUMN c INT;
ALTER TABLE
root@:26257> SELECT t1.*, t2.b AS t2b FROM t AS t1 JOIN t AS t2 ON t1.a = t2.a;
+---+---+------+-----+
| a | b |  c   | t2b |
+---+---+------+-----+
| 1 | 1 | NULL |   1 |
| 2 | 2 | NULL |   2 |
+---+---+------+-----+
(2 rows)
root@:26257> SELECT * FROM tv;
+---+---+------+
| a | b | t2b  |
+---+---+------+
| 1 | 1 | NULL |
| 2 | 2 | NULL |
+---+---+------+
(2 rows)
```

The logic for expanding stars appears to be a little broader in scope than the logic for normalizing table names, so I'm not sure if there's as well-contained of a fix for it as for #9921 (beyond moving to a semantic representation, of course).

Is this the sort of thing that we should block the rollout of views for? Is it ok to leave as a known issue? Is there a fix that could reasonably be put in place for it?

@knz @dt @nvanbenschoten 

Epic CRDB-2410

Jira issue: CRDB-6144

**Comment by a-robinson** [comment]
As briefly discussed in the SQL meeting yesterday, we should "release" views but should either mark the problem as a known issue or not allow creating views with a star-expansion in them.

To avoid potential backwards-compatibility issues down the road, it seems much safer to prevent creation of views with a star-expansion, even if they kind of work now for simple cases, because it could be tough to migrate them properly down the road. I'll send out a PR for that soon if no one disagrees.

**Comment by a-robinson** [comment]
I'm not going to be working on this for a while since it'll depend on some sort of intermediate representation being in place. Unassigning in the meantime.

**Comment by andy-kimball** [comment]
@radu, I think we should remove this error and add some tests. I can't think of a reason why star expansions wouldn't work with the optimizer, can you?

**Comment by andy-kimball** [comment]
I see, those are difficult semantics to implement. The `opt` tree is our semantic representation. It sounds like we'd need to create a serialized representation of the memo we could store in the descriptor. We've talked before about generating serialization/deserialization code for every operator; I guess this is another possible use case for this (the first use case being using opt scalar trees with DistSQL).

**Comment by andy-kimball** [comment]
Also, @knz, do you happen to know how PG solves these problems?

**Comment by knz** [comment]
Yes I do happen to know this.

Basically pg performs semantic analysis, including name resolution and star expansion, on the AST. The AST + structural references are stored in view descriptors.

So for example after:
```
CREATE TABLE t(x INT, y INT);
CREATE VIEW v AS SELECT * FROM t;
```

you will see in `pg_views` that `v` is displayed as `SELECT t.x, t.y FROM t`, which hints at the star expansion.

Further than that, it's the AST and not the _text_ of the query that is stored. For example you will see `FROM t` in `pg_views` in the default display, but then if you do this:

```
SET search_path = pg_catalog; -- remove 'public' which is default
SELECT * FROM pg_views
```

then lo and behold, our view now shows `FROM public.t`.

(If you then work from a different database/catalog, you'll even see the entire db prefix).

This is because pg stores the AST with numeric table/object erferences (OIDs) in the view descriptor, and then pretty-prints those OIDs as the shortest unambiguous name possible given the context.

**Comment by knz** [comment]
FWIW I still have PR #17501 open to address this and the approach in there could still be used with our current architecture.

**Comment by awoods187** [comment]
I ran into this today testing materialized views.

**Comment by jordanlewis** [comment]
Could we do a quicker fix for this than redoing our whole view implementation to use column ids (#17501)? I'm thinking that we could expand the star eagerly, at view creation time, and serialize the view with columns expanded. That seems to be the same thing as what Postgres does, and I don't see why we couldn't do it today.

**Comment by RaduBerinde** [comment]
Interesting, that doesn't sound too hard.. I put it on the radar for 21.1 for now.

**Comment by martinrode** [comment]
I want to +1 this.

It is painful to have to unfold broader tables in expressions.

**Comment by rytaft** [comment]
cc @kevin-v-ngo and @awoods187 for visibility into possible future feature work

**Comment by ajwerner** [comment]
For anybody who finds themselves here, have a look at https://github.com/cockroachdb/cockroach/issues/10083#issuecomment-873204773 which has some nice context.

**Comment by simplenotezy** [comment]
Not sure why, but when we're trying to create a view like so:

```sql
CREATE OR REPLACE VIEW public. "match" AS
SELECT
	swipe.swiper_id AS user_id,
	swipe.swiped_id AS match_id,
	swipe.swiper_cell AS user_cell,
	swipe.swiped_cell AS match_cell
FROM
	swipe
	JOIN active_users u1 ON u1.id = swipe.swiper_id
	JOIN active_users u2 ON u2.id = swipe.swiped_id
WHERE
	swipe.swiper_like
	AND swipe.swiped_like
	AND swipe.deleted_at IS NULL;
```

We are getting the same error:

```console
Query 1 ERROR: ERROR:  unimplemented: views do not currently support * expressions
HINT:  You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/10028/v22.1
```

Not sure why, since we're not using `SELECT *`

**Comment by chengxiong-ruan** [comment]
Thanks for tracking with this issue. I think I'm running into the same problem when tracking column dependencies for UDFs :) glad I ended up here.

**Comment by vy-ton** [comment]
@chengxiong-ruan will any of the planned UDF work help lift this restriction?

**Comment by ajwerner** [comment]
@rharding6373 can we apply #95710 to fix this issue in exactly the same way?

**Comment by knz** [comment]
> @rharding6373 can we apply https://github.com/cockroachdb/cockroach/pull/95710 to fix this issue in exactly the same way?

I have investigated this further today. No we cannot do it easily.

On the one hand, removing the condition on `inViewDef` makes the CREATE VIEW statement succeed and derive an appropriate number of columns for the view schema.

On the other hand, the moment we do `ALTER ADD COLUMN` on the table underlying the view, the columns don't align and we get an execution panic.

So handling views properly here means we need to rewrite the `*` in the SQL stored in the view descriptor, or use a different (binary) encoding of the AST.

I am going to check next whether the UDFs behave properly in that case too.

**Comment by ajwerner** [comment]
I do not understand. Postgres eagerly binds a `*` to the set of columns which exist at the time the view was defined. The view is not affected by adding a column to the underlying table. We ought to do what postgres does and not what was initially described.

**Comment by ajwerner** [comment]
> I am going to check next whether the UDFs behave properly in that case too.

The story for UDFs is more interesting. Postgres has two approaches to UDFs. The "legacy" textual definition, we sometimes called late-bound (`AS $$ ... $$`), and the modern early-bound format `BEGIN ATOMIC; ...; END`. Cockroach, for better or for worse, chose to always early-bound UDF bodies, even when using the legacy syntax.

I assume what we care about is supporting the `RETURNS TABLE` syntax (note that `SETOF <tablename>` returns a single record value). In the late-bound type, with a statically stated table return type, changing the shape of the table behind the `*` will break the UDF.  In the early-bound format, the star is expanded eagerly like in views and thus will continue to work and will not be affected by the added column.

**Comment by Lewenhaupt** [comment]
According to this thread this should be closed but it appears to still be an issue? Just tried adding a simple view Joining on one other table and selecting main_table.* and it gives me the same error as above and links to this issue.

**Comment by rytaft** [comment]
@Lewenhaupt this issue is only fixed in CockroachDB v23.1.x. Serverless clusters have not yet been upgraded to v23.1.x, but the plan is for this to happen in the next couple of weeks.
